pkg/grpc/server.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2017 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+// Package server provides an implementation of a streaming xDS server
+package server
+
+import (
+	"strconv"
+	"sync/atomic"
+
+	"github.com/envoyproxy/go-control-plane/api"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	context "golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Resource types in xDS v2
+const (
+	typePrefix   = "type.googleapis.com/envoy.api.v2."
+	EndpointType = typePrefix + "LbEndpoint"
+	ClusterType  = typePrefix + "Cluster"
+	RouteType    = typePrefix + "Route"
+	ListenerType = typePrefix + "Listener"
+)
+
+// Server is a collection of handlers for streaming discovery requests
+type Server interface {
+	// Register the handlers in a gRPC server
+	Register(*grpc.Server)
+}
+
+// NewServer creates handlers from a config watcher
+func NewServer(config cache.ConfigWatcher) Server {
```

Comment:
Why not just make the rpc.Server time public and return it rather than using an interface?

Reply:
The server is not done, I want to use interface to encapsulate the future stuff to be added.